### PR TITLE
Persistent tokens

### PR DESCRIPTION
**Pull request by ghophp**
This changes allow the user to keep logged in even when a new version of the software get deployed. Also, it makes the oauth2-provider a step closer to stateless.

```securesocial``` is using the ```CacheAPI``` from play.. that uses ```Ehcache``` that claims itself as a reliable solution. It is not that simple to replace this cacheservice with a key hash database, so we will split this in another story at the backlog, that will be done after the pilot.

Also the email tokens are being persisted, so even if some deployment happen, the mail tokens are stored at the current riak cluster.. so no "limbo" email confirmation.

If you have oauth2-provider locally, you need to make some changes at your application.conf, please follow the changes of the application-example.conf to do so.

### CODE REVIEW

File: app/com/yetu/oauth2provider/oauth2/handlers/AuthorizationHandler.scala [hunk 1]
```diff
@@ -37,8 +37,8 @@ class AuthorizationHandler(authAccessService: IAuthCodeAccessTokenService,
         logger.debug(s"id = ${oauthClient.clientId} , secret = ${oauthClient.clientSecret}, grantType = ${oauthClient.grantTypes}")
 
         val validClientId = oauthClient.clientId == clientCredential.clientId
-        val validGrantType = oauthClient.grantTypes.map(grantList => grantList.contains(grantType)).getOrElse(false)
-        val validSecret = clientCredential.clientSecret.map(_ == oauthClient.clientSecret).getOrElse(false)
+        val validGrantType = oauthClient.grantTypes.exists(grantList => grantList.contains(grantType))
+        val validSecret = clientCredential.clientSecret.contains(oauthClient.clientSecret)
```
Comment (jschaul) [inline app/com/yetu/oauth2provider/oauth2/handlers/AuthorizationHandler.scala]: comparing for equality and comparing whether one is a substring of the other is not quite the same thing. Why did you change this code?

File: app/com/yetu/oauth2provider/oauth2/handlers/AuthorizationHandler.scala [hunk 2]
```diff
@@ -106,43 +99,27 @@ class AuthorizationHandler(authAccessService: IAuthCodeAccessTokenService,
     Future.successful(loggedIn)
   }
 
-  def getStoredAccessToken(authInfo: AuthInformation) = {
+  def getStoredAccessToken(authInfo: AuthInfo[YetuUser]) = {
     logger.warn("...getStoredAccessToken :: NOT_IMPLEMENTED")
```
Comment (jschaul) [inline app/com/yetu/oauth2provider/oauth2/handlers/AuthorizationHandler.scala]: Can we see where this method is used and see if we can implement it, now that we have some persistence? This however does the reverse lookup.

Comment (ghophp) [inline app/com/yetu/oauth2provider/oauth2/handlers/AuthorizationHandler.scala]: Good.. let's take a look..

File: app/com/yetu/oauth2provider/services/data/RiakAuthCodeAccessTokens.scala [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+
+  implicit def formatAuthInfo[T: Format]: Format[AuthInfo[T]] =
+    ((__ \ "user").format[T] and
+      (__ \ "clientId").formatNullable[String] and
+      (__ \ "scope").formatNullable[String] and
+      (__ \ "redirectUri").formatNullable[String]
+    )(AuthInfo.apply, unlift(AuthInfo.unapply))
+
+  import scala.concurrent.ExecutionContext.Implicits.global
+  val logger = Logger(this.getClass)
```
Comment (jschaul) [inline app/com/yetu/oauth2provider/services/data/RiakAuthCodeAccessTokens.scala]: we can remove this line and instead extend `NamedLogger` for consistency

File: app/com/yetu/oauth2provider/oauth2/handlers/AuthorizationHandler.scala (hunk 1 above)
Comment (ghophp) [inline app/com/yetu/oauth2provider/oauth2/handlers/AuthorizationHandler.scala]: It looks better that way don't you think? Is not the same, but it will work properly in this case, don't you think?

File: app/com/yetu/oauth2provider/services/data/RiakAuthCodeAccessTokens.scala (hunk 3 above)
Comment (ghophp) [inline app/com/yetu/oauth2provider/services/data/RiakAuthCodeAccessTokens.scala]: :+1:

File: conf/application-example.conf [hunk 4]
```diff
@@ -35,4 +35,10 @@ securesocial{
...
   ssl=false
-}
\ No newline at end of file
+}
+
+rabbitmq {
+  host = "rabbit000.dev.yetu.me"
+}
+
+riak.production.host = "riak000.dev.yetu.me"
```
Comment (jschaul) [inline conf/application-example.conf]: add `riak.test.host`

File: test/com/yetu/oauth2provider/services/AuthCodeAccessTokensSpec.scala [hunk 5]
```diff
@@ -0,0 +1,7 @@
+package com.yetu.oauth2provider.services
+
+import com.yetu.oauth2provider.base.BaseSpec
+
+class AuthCodeAccessTokensSpec extends BaseSpec {
+
+}
```
Comment (jschaul) [inline test/com/yetu/oauth2provider/services/AuthCodeAccessTokensSpec.scala]: remove this class?

File: test/com/yetu/oauth2provider/services/RiakHandlerITSpec.scala [hunk 6]
```diff
@@ -0,0 +1,92 @@
...
+import scala.concurrent.Future
+import scalaoauth2.provider.{ AuthInfo, AccessToken }
+
+import scala.concurrent.ExecutionContext.Implicits.global
+
+import org.scalatest.time.{ Millis, Seconds, Span }
+
+class RiakHandlerITSpec extends BaseIntegrationSpec with ScalaFutures with AsyncAssertions {
+
+  //with ParallelTestExecution
```
Comment (jschaul) [inline test/com/yetu/oauth2provider/services/RiakHandlerITSpec.scala]: remove this line

File: conf/application-example.conf (hunk 4 above)
Comment (ghophp) [inline conf/application-example.conf]: it is not necessary remember? the test line was just necessary because I was running from my IDE without -D parameter.. but if you run from sbt commands, that are the recommended way.. it isn't necessary..

File: test/com/yetu/oauth2provider/services/AuthCodeAccessTokensSpec.scala (hunk 5 above)
Comment (ghophp) [inline test/com/yetu/oauth2provider/services/AuthCodeAccessTokensSpec.scala]: :+1:

File: app/com/yetu/oauth2provider/services/data/RiakCacheService.scala [hunk 7]
```diff
@@ -0,0 +1,27 @@
...
+import com.yetu.oauth2provider.data.riak.RiakConnection
+import com.yetu.oauth2provider.oauth2.models.YetuUser
+import com.yetu.oauth2provider.utils.NamedLogger
+import play.api.libs.json.Json
+import securesocial.core.services.CacheService
+
+import scala.concurrent.ExecutionContext.Implicits.global
+import scala.concurrent.Future
+
+class RiakCacheService(riakConnection: RiakConnection) extends CacheService with NamedLogger {
```
Comment (ghophp) [inline app/com/yetu/oauth2provider/services/data/RiakCacheService.scala]: I will remove this..

File: app/com/yetu/oauth2provider/controllers/OAuth2Validation.scala [hunk 8]
```diff
@@ -2,17 +2,15 @@ package com.yetu.oauth2provider
...
-import com.yetu.oauth2resource.model.ValidationResponse
-import play.api.libs.json.{ JsValue, Json }
 import play.api.mvc.Action
 
 import scala.concurrent.Future
-import scalaoauth2.provider.AuthInfo
 
 class OAuth2Validation(validationService: ValidationService, authorizationHandler: AuthorizationHandler) extends OAuth2Controller {
 
+  import scala.concurrent.ExecutionContext.Implicits.global
```
Comment (larsblumberg) [inline app/com/yetu/oauth2provider/controllers/OAuth2Validation.scala]: This import is not used

File: app/com/yetu/oauth2provider/oauth2/services/AuthorizeService.scala [hunk 9]
```diff
@@ -155,11 +153,11 @@ class AuthorizeService(authAccessService: IAuthCodeAccessTokenService,
   def handleClientPermissions(client: OAuth2Client, authorizeRequest: AuthorizeRequest, user: YetuUser): Result = {
     val clientPermission: Option[ClientPermission] = permissionService.findPermission(user.identityId.userId, client.clientId)
     clientPermission match {
-      case None => {
+      case None =>
         //TODO: FIXME
         Ok("TODO: permissions form should be here!")
-        //Ok(com.yetu.oauth2provider.views.html.permissions(permissionsForm, client.clientName, Some(client.clientId), authorizeRequest.redirectUri, Some(authorizeRequest.state)))
-      }
+      //Ok(com.yetu.oauth2provider.views.html.permissions(permissionsForm, client.clientName, Some(client.clientId), authorizeRequest.redirectUri, Some(authorizeRequest.state)))
```
Comment (larsblumberg) [inline app/com/yetu/oauth2provider/oauth2/services/AuthorizeService.scala]: Do you still need this commented code? If so, can you add a `TODO` comment?

File: app/com/yetu/oauth2provider/services/data/MemoryAuthCodeAccessTokens.scala [hunk 10]
```diff
@@ -12,117 +14,56 @@ import scalaoauth2.provider.AccessToken
...
   }
 
-  def saveAccessTokenToUser(accessToken: AccessToken, authInfo: AuthInformation) = {
-    logger.debug(s"saveAccessTokenToUser accessToken: $accessToken")
-    logger.debug(s"saveAccessTokenToUser authInfo: $authInfo")
-    accessTokensWithUser += (accessToken -> authInfo)
+  def findAuthInfoByAccessToken(token: String) = {
+    val authInfo: Option[AuthInfo[YetuUser]] = accessTokensToAuthInfo.get(token)
+    logger.debug(s"findUserByAccessToken token: $token")
+    logger.debug(s"findUserByAccessToken authInfo: $authInfo")
```
Comment (larsblumberg) [inline app/com/yetu/oauth2provider/services/data/MemoryAuthCodeAccessTokens.scala]: Previously you logged `token` and `authInfo` in one single line. You might want to use the same pattern.
